pkg/placement/placement-decision.go

```diff
@@ -0,0 +1,144 @@
+	return nil
+}
+
+// updateOrCreatePlacementDecision updates or creates a placement-decision object in the cluster.
+// If the object already exists, it is updated. Otherwise, it is created.
+func (c *Controller) updateOrCreatePlacementDecision(pd *v1alpha1.PlacementDecision,
+	placementDecisionSpec *v1alpha1.PlacementDecisionSpec) error {
+	// use the passed placement decision and set its spec
+	pd.Spec = *placementDecisionSpec
+
+	// set owner reference
+	//	get owner placement object
+	placement, err := c.getPlacementByName(pd.GetName())
+	if err != nil {
+		return fmt.Errorf("failed to get placement %s: %w", pd.GetName(), err)
+	}
+
+	setOwnerReference(placement, pd)
+
+	// update or create placement decision
+	unstructuredPlacementDecision, err := placementDecisionToUnstructuredObject(pd)
+	if err != nil {
+		return fmt.Errorf("failed to update or create placement decision: %w", err)
+	}
+
+	_, err = c.dynamicClient.Resource(schema.GroupVersionResource{
+		Group:    v1alpha1.SchemeGroupVersion.Group,
+		Version:  pd.GetObjectKind().GroupVersionKind().Version,
+		Resource: util.PlacementDecisionResource,
+	}).Update(context.Background(), unstructuredPlacementDecision, metav1.UpdateOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			_, err = c.dynamicClient.Resource(schema.GroupVersionResource{
+				Group:    v1alpha1.SchemeGroupVersion.Group,
+				Version:  pd.GetObjectKind().GroupVersionKind().Version,
+				Resource: util.PlacementDecisionResource,
+			}).Create(context.Background(), unstructuredPlacementDecision, metav1.CreateOptions{})
+			if err != nil {
+				return fmt.Errorf("failed to create placement decision: %w", err)
+			}
+		} else {
+			return fmt.Errorf("failed to update placement decision: %w", err)
+		}
+	}
+
+	c.logger.Info("updated placement decision", "name", pd.GetName())
+	return nil
+}
+
+func unstructuredObjectToPlacementDecision(unstructuredObj *unstructured.Unstructured) (*v1alpha1.PlacementDecision, error) {
+	var placementDecision *v1alpha1.PlacementDecision
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredObj.UnstructuredContent(),
+		&placementDecision); err != nil {
+		return nil, fmt.Errorf("failed to convert Unstructured to PlacementDecision: %w", err)
+	}
+
+	return placementDecision, nil
+}
+
+func placementDecisionToUnstructuredObject(placementDecision *v1alpha1.PlacementDecision) (*unstructured.Unstructured, error) {
+	innerObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(placementDecision)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert PlacementDecision to Unstructured: %w", err)
+	}
+
+	return &unstructured.Unstructured{
+		Object: innerObj,
+	}, nil
+}
+
+func setOwnerReference(ownerObj runtime.Object, placementDecision *v1alpha1.PlacementDecision) {
+	// set the placement as the only owner-reference of the placement-decision
+	ownerRef := metav1.NewControllerRef(ownerObj.(metav1.Object), ownerObj.GetObjectKind().GroupVersionKind())
```

Comment:
Note to self: If the garbage collector is going to delete the PlacementDecision then this controller does not need to. Does it?

Reply:
Indeed. From what I understood, setting controller to true means setting different owner-references would be blocked, correct?

So when a placement is deleted, the relevant placement-decision is deleted by the garbage collector, and that's it. This controller doesn't have any placement-decision deleting logic, it only deletes the resolution when handling the placement deletion event.

Reply:
I don't know about preventing other OwnerReferences (and you did not qualify that statement regarding whether the others have Controller set).